### PR TITLE
[GLUTEN-1914][CH] Fix mismatch value of complex types

### PR DESCRIPTION
**Pull request by taiyang-li**
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

(Fixes: \#1914) 

Notice: this pr must be merged after https://github.com/Kyligence/ClickHouse/pull/450.

### CODE REVIEW

File: cpp-ch/local-engine/Parser/TypeParser.cpp [hunk 1]
```diff
@@ -146,19 +146,22 @@ DB::DataTypePtr TypeParser::parseType(const substrait::Type & substrait_type, st
...
     {
-        DB::DataTypes struct_field_types(substrait_type.struct_().types().size());
+        const auto & types = substrait_type.struct_().types();
+        DB::DataTypes struct_field_types(types.size());
         DB::Strings struct_field_names;
-        for (int i = 0; i < static_cast<int>(struct_field_types.size()); ++i)
+        for (int i = 0; i < types.size(); ++i)
         {
+            struct_field_types[i] = parseType(types[i], field_names);
+
```
Comment (lgbo-ustc) [inline cpp-ch/local-engine/Parser/TypeParser.cpp:155]: Why move this in front of 'if'

Comment (taiyang-li) [reply cpp-ch/local-engine/Parser/TypeParser.cpp:155]: Initialize the two var in order as they are passed to `ch_type = std::make_shared<DB::DataTypeTuple>(struct_field_types, struct_field_names);`

File: cpp-ch/local-engine/Storages/ch_parquet/OptimizedArrowColumnToCHColumn.cpp [hunk 2]
```diff
@@ -1,69 +1,68 @@
...
+#        include <Columns/ColumnUnique.h>
+#        include <Columns/ColumnsNumber.h>
+#        include <Core/Block.h>
+#        include <DataTypes/DataTypeArray.h>
+#        include <DataTypes/DataTypeDate.h>
+#        include <DataTypes/DataTypeDate32.h>
+#        include <DataTypes/DataTypeDateTime64.h>
+#        include <DataTypes/DataTypeFactory.h>
+#        include <DataTypes/DataTypeLowCardinality.h>
+#        include <DataTypes/DataTypeMap.h>
```
Comment (lgbo-ustc) [inline cpp-ch/local-engine/Storages/ch_parquet/OptimizedArrowColumnToCHColumn.cpp:22]: Remove this spaces?

File: cpp-ch/local-engine/Storages/ch_parquet/OptimizedArrowColumnToCHColumn.h [hunk 3]
```diff
@@ -4,7 +4,6 @@
 #include <Common/Config.h>
 
 #if USE_LOCAL_FORMATS
-// clang-format off
 #if USE_ARROW || USE_ORC || USE_PARQUET
```
Comment (lgbo-ustc) [inline cpp-ch/local-engine/Storages/ch_parquet/OptimizedArrowColumnToCHColumn.h:7]: remove the `clang-format off` can work now?

File: cpp-ch/local-engine/Storages/ch_parquet/OptimizedArrowColumnToCHColumn.cpp (hunk 2 above)
Comment (taiyang-li) [reply cpp-ch/local-engine/Storages/ch_parquet/OptimizedArrowColumnToCHColumn.cpp:22]: it is the output of clang-format.

File: cpp-ch/local-engine/Storages/ch_parquet/OptimizedArrowColumnToCHColumn.h (hunk 3 above)
Comment (taiyang-li) [reply cpp-ch/local-engine/Storages/ch_parquet/OptimizedArrowColumnToCHColumn.h:7]: it does not influence building .

File: cpp-ch/local-engine/Parser/TypeParser.cpp (hunk 1 above)
Comment (taiyang-li) [reply cpp-ch/local-engine/Parser/TypeParser.cpp]: I am wrong.. it is reverted now.

**Comment by taiyang-li** [comment]
@baibaichen @liuneng1994 This fix pr relies on a patch which modifies clickhouse kernel code, which I'm afraid CH community would refuse it. Can we just merge it to https://github.com/kyligence/clickhouse directly 


[arrow2ch.patch](https://github.com/oap-project/gluten/files/12077077/arrow2ch.patch)

**Comment by taiyang-li** [comment]
0: jdbc:hive2://localhost:10000/> CREATE TABLE t(
. . . . . . . . . . . . . . . . >   id INT,
. . . . . . . . . . . . . . . . >   info STRUCT<name:STRING, age:INT>,
. . . . . . . . . . . . . . . . >   data MAP<STRING, INT>,
. . . . . . . . . . . . . . . . >   values ARRAY<INT>
. . . . . . . . . . . . . . . . > ) stored as parquet;
+---------+
| Result  |
+---------+
+---------+
No rows selected (3.582 seconds)
0: jdbc:hive2://localhost:10000/> 
0: jdbc:hive2://localhost:10000/> INSERT overwrite t VALUES
. . . . . . . . . . . . . . . . >   (1, struct('John', 25), map('A', 10, 'B', 20), array(1.0, 2.0, 3.0)),
. . . . . . . . . . . . . . . . >   (2, struct('Alice', 30), map('C', 15, 'D', 25), array(4.0, 5.0, 6.0)),
. . . . . . . . . . . . . . . . >   (3, struct('Bob', 35), map('E', 12, 'F', 18), array(7.0, 8.0, 9.0)),
. . . . . . . . . . . . . . . . >   (4, struct('Jane', 40), map('G', 22, 'H', 30), array(10.0, 11.0, 12.0)),
. . . . . . . . . . . . . . . . >   (5, struct('Kate', 45), map('I', 17, 'J', 28), array(13.0, 14.0, 15.0)),
. . . . . . . . . . . . . . . . >   (6, null, null, null),
. . . . . . . . . . . . . . . . >   (7, struct('Tank', 20), map('X', null, 'Y', null), array(1.0, 2.0, 3.0));
+---------+
| Result  |
+---------+
+---------+
No rows selected (6.366 seconds)
0: jdbc:hive2://localhost:10000/> 
0: jdbc:hive2://localhost:10000/> 
0: jdbc:hive2://localhost:10000/> 
0: jdbc:hive2://localhost:10000/> 
0: jdbc:hive2://localhost:10000/> select * from t; 
+-----+----------------------------+----------------------+-------------+
| id  |            info            |         data         |   values    |
+-----+----------------------------+----------------------+-------------+
| 1   | {"name":"John","age":25}   | {"A":10,"B":20}      | [1,2,3]     |
| 2   | {"name":"Alice","age":30}  | {"C":15,"D":25}      | [4,5,6]     |
| 5   | {"name":"Kate","age":45}   | {"I":17,"J":28}      | [13,14,15]  |
| 6   | NULL                       | NULL                 | NULL        |
| 7   | {"name":"Tank","age":20}   | {"X":null,"Y":null}  | [1,2,3]     |
| 3   | {"name":"Bob","age":35}    | {"E":12,"F":18}      | [7,8,9]     |
| 4   | {"name":"Jane","age":40}   | {"G":22,"H":30}      | [10,11,12]  |
+-----+----------------------------+----------------------+-------------+

File: cpp-ch/clickhouse.version [hunk 4]
```diff
@@ -1,3 +1,3 @@
 CH_ORG=Kyligence
 CH_BRANCH=rebase_ch/20230717
```
Comment (binmahone) [inline cpp-ch/clickhouse.version:2]: 这里branch name应该和commit 同步修改

**Review by binmahone — CHANGES_REQUESTED** [review]

**Review by binmahone — APPROVED** [review]

**Pull request merged**